compiler-rt/lib/scudo/standalone/stack_depot.h

```diff
@@ -136,6 +136,14 @@ class StackDepot {
   u64 operator[](uptr RingPos) const {
     return atomic_load_relaxed(&Ring[RingPos & RingMask]);
   }
+
+  void disable() NO_THREAD_SAFETY_ANALYSIS {
+    RingEndMu.lock();
+  }
+
+  void enable() NO_THREAD_SAFETY_ANALYSIS {
+    RingEndMu.unlock();
+  }
```

Comment:
Can you share more details about the deadlock? From where the `RingEndMu` is used (only in the `insert()` above and it seems to me that it's not disabling the StackDepot. For example, `find()` is still working when we call `disable()`. 
I'm wondering if we want a different lock to guard all the operations (we may not be able to lock in the user because it passes the address of `Depot`)

Reply:
Right, StackDepot is not actually being disabled here. The idea is grab all allocator locks in a pthread_atfork before the fork, and release them after. This allows malloc to be used in a fork child of a multithreaded process, which is expressly forbidden by the standard, but very widely used. For example, Android's `init` uses `std::string` after fork when spawning services, for example in `android::init::EnterNamespaces`.

Any lock that is necessary to serve an allocator call must be handled this way. Otherwise there is a possibility that the lock is held during the call to fork, which results in it being held forever in the child process, and the next operation that needs it deadlocks.

Reply:
I see. Thanks for the explanation.

For the use case in `release.h`, they are supposed to be acquired after the region mutexes of primary allocator. Which means once you disable the primary allocator, those mutexes won't be held. Given the constraint of thread safety analysis, we don't have an easy way to express that but I can add some runtime debug checks to ensure that assumption.

For the `StackDepot`, even the names `enable/disable` seem to be slightly misleading here but I think it's fine according to the context. Can we add some short comment to the `enable/disable` ?

Reply:
Great point about the BufferPool lock, removed it from the pull request.
Added a comment in StackDepot.